File: SoccerAgent/src/test/java/onflows/calculation/structure/flows/AssociativePlainFlow.java
package onflows.calculation.structure.flows;

import onflows.worldmodel.Relation;
import onflows.worldmodel.WorldObject;

import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Supplier;

/**
 * Created by Sergey on 13.09.2016.
 */
public class AssociativePlainFlow
        <D extends WorldObject, R extends WorldObject, S extends Relation<D, R>>
        extends RelationFlow<S> {

    private final BiFunction<D, R, S> doStaff;
    private Supplier<List<D>> domainFlow;
    private Supplier<List<R>> rangeFlow;
    private List<S> result = new ArrayList<>();

    public AssociativePlainFlow(BiFunction<D, R, S> doStaff) {
        this.doStaff = doStaff;
    }

    @Override
    public void calculate() {
        if (checkLInFlow(domainFlow.get()) && checkLInFlow(rangeFlow.get())) {
            for (D d : domainFlow.get()) {
                for (R r : rangeFlow.get()) {
                    result.add(doStaff.apply(d, r));
                }
            }
        }
    }

    public List<S> getResult() {
        return result;
    }

    public void setDomainFlow(Supplier<List<D>> domainFlow) {
        this.domainFlow = domainFlow;
    }

    public void setRangeFlow(Supplier<List<R>> rangeFlow) {
        this.rangeFlow = rangeFlow;
    }
}

File: ProtoSASOnScala/src/main/scala/dd/sas/pipeline/calculation/structures/flows/associativeflows/AssociativePlainFlow.scala
package dd.sas.pipeline.calculation.structures.flows.associativeflows

import java.util.function.BiFunction

import dd.sas.pipeline.calculation.result.Answer
import dd.sas.pipeline.calculation.structures.flows.Flow
import dd.sas.pipeline.calculation.structures.flows.generativeflows.ObjectFlow
import dd.sas.pipeline.calculation.structures.nodes.Node
import dd.sas.pipeline.worldmodel.{Relation, WorldObject}

/**
  * Created by Sergey on 24.09.2016.
  */
class AssociativePlainFlow[D <: WorldObject, R <: WorldObject, S <: Relation[D, R]]
(node: Node, expression: BiFunction[D, R, S]) extends RelationFlow[D, R, S](node) {

  private var domainFlow: ObjectFlow[D] = _
  private var rangeFlow: ObjectFlow[R] = _
  private var result: List[S] = _

  //need this method cuz we derive node when there are some flows, but not all of them
  private var customProcessVar: () => Unit = () => {
    if(domainFlow != null && rangeFlow != null){
      customProcessVar = customProcessActivated
      customProcessActivated()
    }
  }

  override def apply(): List[S] = {
    result
  }

  def customProcessActivated(): Unit = {
    result = for {
      d <- domainFlow()
      r <- rangeFlow()
    } yield expression(d, r)
    resubcribe()
  }

  override def customProcess(): Unit ={
    customProcessVar()
  }


  def setDomainFlow(domainFlow: Flow): Unit = {
    this.domainFlow = domainFlow.asInstanceOf[ObjectFlow[D]]
    donors.put(domainFlow, this.setDomainFlow)
    setToBeUpdated()
  }

  def setRangeFlow(rangeFlow: Flow): Unit = {
    this.rangeFlow = rangeFlow.asInstanceOf[ObjectFlow[R]]
    donors.put(rangeFlow, this.setRangeFlow)
    setToBeUpdated()
  }

}
